pkg/controller/source/source_test.go

```diff
@@ -251,11 +253,194 @@ var _ = Describe("Source", func() {
 	})
 
 	Describe("ChannelSource", func() {
-		It("TODO(community): implement this", func(done Done) {
-			instance := source.ChannelSource(make(chan event.GenericEvent))
-			instance.Start(nil, nil)
+		var stop chan struct{}
+		var ch chan event.GenericEvent
 
-			close(done)
+		BeforeEach(func() {
+			stop = make(chan struct{})
+			ch = make(chan event.GenericEvent)
+		})
+
+		AfterEach(func() {
+			close(stop)
+			close(ch)
+		})
+
+		Context("for a source", func() {
+			It("should provide a GenericEvent", func(done Done) {
+				ch := make(chan event.GenericEvent)
+				c := make(chan struct{})
+				p := &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "bar"},
+				}
+				evt := event.GenericEvent{
+					Object: p,
+					Meta:   p,
+				}
+
+				q := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "test")
+				instance := &source.ChannelSource{Source: ch}
+				inject.DoStop(stop, instance)
+				err := instance.Start(eventhandler.Funcs{
+					CreateFunc: func(workqueue.RateLimitingInterface, event.CreateEvent) {
+						defer GinkgoRecover()
+						Fail("Unexpected CreateEvent")
+					},
+					UpdateFunc: func(workqueue.RateLimitingInterface, event.UpdateEvent) {
+						defer GinkgoRecover()
+						Fail("Unexpected UpdateEvent")
+					},
+					DeleteFunc: func(workqueue.RateLimitingInterface, event.DeleteEvent) {
+						defer GinkgoRecover()
+						Fail("Unexpected DeleteEvent")
+					},
+					GenericFunc: func(q2 workqueue.RateLimitingInterface, evt event.GenericEvent) {
+						defer GinkgoRecover()
+						Expect(q2).To(Equal(q))
+						Expect(evt.Meta).To(Equal(p))
+						Expect(evt.Object).To(Equal(p))
+						close(c)
+					},
+				}, q)
+				Expect(err).NotTo(HaveOccurred())
+
+				ch <- evt
+				<-c
+				close(done)
+			})
+			It("should block if exceed buffer size", func(done Done) {
```

Comment:
I think we care less that it blocks, and more that it doesn't blow up.  I think the main thing we want to test is that we can write 3 events to the source channel while the eventhandler is blocked, then unblock the eventhandler, and all the 3 events will get processed.